katjab rudimentary Final Project Proposal

bkat70 opened the pull request.

bkat70 committed:
updated file to include my rudimentary Final Project Proposal

bkat70 committed:
updated Final Project Proposal

jamdon commented:
I think you could do either of your ideas. While we haven't gotten into what Backbone is yet, it's not really necessary to understand what it is as it's really just a toolset to help you organize code. You should do whichever project you're most interested in and feel confident that you can complete given the time constraints you'll have.

bkat70 committed:
added final project files

bkat70 committed:
removed obsolete server routes and updated readme

bkat70 committed:
updated all sorts of files

jamdon commented on public/index.html:
```diff
@@ -0,0 +1,33 @@
+		<img id="bkat" src="img/bkat.jpg" alt="bKat" style="height:50px;"><br><br>
+	</div>
+	<div class="center" id="choices">
+		<a href="/api/houses" class="button" id="allHouses">View ALL Homes for Sale</a>	
```
I would avoid confusing your api routes with your client-side routes. A client-side route should be able to load the application up so that it displays in the same state it appeared in if you had loaded the application from the root then navigated to this specific route. In this case, you're using the routes you've defined in your express router, which are really api endpoints meant to be used for ajax calls to retrieve or update data. If you appRouter.navigate to /api/houses, then refresh the browser, it simply calls your router endpoint and tries to display the resulting data in the browser. Instead, use a route like /houses. To facilitate the ability to load routes like this and not have the server complain, look at how I used the spa-server-route in the food-truck-app index.js. You're welcome to use the same module to make this work for your application. The key takeaway is that your client-side routing needs to work if you load the routes to start with - this is one of the primary responsibilities of the Backbone Router.

jamdon commented:
The application is looking good. There's one issue with the way your routes are set up, and I've left a comment inline about it. Let me know if what I'm saying isn't clear.

bkat70 commented:
hi James,

did I accidentally post my project in the wrong place? Sorry, if I did but
thanks for checking it out! It's not quite done (haven't linted it yet
etc.).

I am not certain I understand all that you were saying. I got that somehow
my client router isn't called/used but I don't know why not? I thought I
copied your set-up :)

I did not use the spaRoute because we've never talked about it before and
therefore I wasn't convinced it was strictly necessary. If I implement this
route, will the problem be solved? Why?

thank you!
katja

On Tue, Jun 14, 2016 at 11:06 AM, jamdon notifications@github.com wrote:

> The application is looking good. There's one issue with the way your
> routes are set up, and I've left a comment inline about it. Let me know if
> what I'm saying isn't clear.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/JSCRIPT300-Spring2016/Final_Project/pull/2#issuecomment-225966156,
> or mute the thread
> https://github.com/notifications/unsubscribe/ARbciFmuQE561WJL5WPAQEuZsOYihdsDks5qLu22gaJpZM4IRjAv
> .

bkat70 commented:
hi James,

I don't see a 'path' node module in the FoodTruck API package.json, yet
it's required in in index.js. Is it a 'hidden' module of some sorts?

thanks!
katja

On Tue, Jun 14, 2016 at 7:04 PM, Katja Borchert katjab@uw.edu wrote:

> hi James,
> 
> did I accidentally post my project in the wrong place? Sorry, if I did but
> thanks for checking it out! It's not quite done (haven't linted it yet
> etc.).
> 
> I am not certain I understand all that you were saying. I got that somehow
> my client router isn't called/used but I don't know why not? I thought I
> copied your set-up :)
> 
> I did not use the spaRoute because we've never talked about it before and
> therefore I wasn't convinced it was strictly necessary. If I implement this
> route, will the problem be solved? Why?
> 
> thank you!
> katja
> 
> On Tue, Jun 14, 2016 at 11:06 AM, jamdon notifications@github.com wrote:
> 
> > The application is looking good. There's one issue with the way your
> > routes are set up, and I've left a comment inline about it. Let me know if
> > what I'm saying isn't clear.
> > 
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > https://github.com/JSCRIPT300-Spring2016/Final_Project/pull/2#issuecomment-225966156,
> > or mute the thread
> > https://github.com/notifications/unsubscribe/ARbciFmuQE561WJL5WPAQEuZsOYihdsDks5qLu22gaJpZM4IRjAv
> > .

jamdon commented:
Your Backbone router is handling `/`, `/houses`, and `/houses/:id`, but you're linking to `/api/houses` and `/api/houses/:id` in your application, which are your Express server routes. You want to link to the same routes in your Backbone router. If you've updated the url to `/houses`, then you refresh the page, your server is going to try to serve up `/houses`, but it won't have any routes defined for it. This is why you use something like the spa-server-route, so that for routes your server hasn't defined - it will just serve up index.html and run your backbone application again. Your Backbone router will handle the route action instead. Does that make sense?

bkat70 commented:
hi James,

I added the SpaRoute code and changed the href of my buttons in index.html
and it seems to work now (well, at least as far as I can tell). I also
npm-ed in "path" but I still don't know if it's necessary? The truck API
seems to work without it being explicitly installed? I'm just curious.

thanks for all your help!
katja

On Tue, Jun 14, 2016 at 8:47 PM, jamdon notifications@github.com wrote:

> Your Backbone router is handling /, /houses, and /houses/:id, but you're
> linking to /api/houses and /api/houses/:id in your application, which are
> your Express server routes. You want to link to the same routes in your
> Backbone router. If you've updated the url to /houses, then you refresh
> the page, your server is going to try to serve up /houses, but it won't
> have any routes defined for it. This is why you use something like the
> spa-server-route, so that for routes your server hasn't defined - it will
> just serve up index.html and run your backbone application again. Your
> Backbone router will handle the route action instead. Does that make sense?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/JSCRIPT300-Spring2016/Final_Project/pull/2#issuecomment-226081081,
> or mute the thread
> https://github.com/notifications/unsubscribe/ARbciNdMn4E-gdDxIKl1pFZQXClxk_7Nks5qL3XBgaJpZM4IRjAv
> .

jamdon commented:
The path module will resolve paths so they are platform independent, making the code more portable. Your code will work without it, it just makes it more robust.

bkat70 committed:
updated finalProject

bkat70 committed:
linted AppView.js

bkat70 committed:
forgot to grunt one last time

bkat70 commented:
hi James,

I made a pull request (actually two but I noticed that I had forgotten to
grunt one last time before the pull request). If there is something not
working (if you even have time to check it), let me know.

Again, thank you so much for all your help!
cheers,
katja

On Tue, Jun 14, 2016 at 10:00 PM, jamdon notifications@github.com wrote:

> The path module will resolve paths so they are platform independent,
> making the code more portable. Your code will work without it, it just
> makes it more robust.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/JSCRIPT300-Spring2016/Final_Project/pull/2#issuecomment-226088344,
> or mute the thread
> https://github.com/notifications/unsubscribe/ARbciPTdLEUJ9Eo5KFSUwq27WmaTv-ipks5qL4bygaJpZM4IRjAv
> .

jamdon commented on public/js/controllers/AppController.js:
```diff
@@ -0,0 +1,68 @@
+  var model = options.model;
+  jqxhr.done(function () {
+    //console.log("id ", model.id);
+    model = houses.findWhere({ _id: model.id });
```
When this is called from the showHouse method in your appRouter (which happens if you load the `houses/:id` route from the server), you're passing in an id on the options object. That means model will be undefined, you'll get a script error, and the rest of the application will stop working. To fix this, I recommend referencing the id from the options object, then passing the id in your `handleClick` method inside the HouseItemView.

jamdon commented:
There's a remaining issue with loading one of your routes from the server and I left a comment inline about it.

bkat70 commented:
hi James,

thanks for letting me know.

However, I have really no idea how to fix this: "_To fix this, I recommend
referencing the id from the options object, then passing the id in
your handleClick_\* method inside the HouseItemView*." I also don't quite
understand the problem: if I click on the link for the individual house,
the individual house is presented and the application does not freeze. So,
I am at loss at what to fix exactly.

_AppController.js:_

//change to id as it's the id that is handed in in router
function showHouse(options) {
console.log(options);
console.log(options.model.id);
  var model = options.model;
  jqxhr.done(function () {
    //console.log("id ", model.id);
    model = houses.findWhere({ _id: model.id });
    //console.log("find where model ", model);
    resultsView.showHouse(model);
  });
}

If I console.log(options), I get an option object; options.model.id give me
the id of the model. Is there something else I am supposed to do?

_HouseItemView.js:_

  handleClick: function (ev) {
    ev.preventDefault();
    var path  = ev.currentTarget.href.replace(location.origin, '');
    vent.trigger('house:selected', { model: this.model, path: path });
  },

How do I pass an id in here and what would I do with it once in the
function?

  handleClick: function (ev, id) {
    ev.preventDefault();
    var path  = ev.currentTarget.href.replace(location.origin, '');
    vent.trigger('house:selected', { model: this.model, path: path });
  },

So, if you have any more words wisdom let me know. I'd like to correct it
but I just don't understand your suggested fix.

thank you!
katja

On Wed, Jun 15, 2016 at 11:02 PM, jamdon notifications@github.com wrote:

> There's a remaining issue with loading one of your routes from the server
> and I left a comment inline about it.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/JSCRIPT300-Spring2016/Final_Project/pull/2#issuecomment-226395542,
> or mute the thread
> https://github.com/notifications/unsubscribe/ARbciIknp-hv_EuB2S2rs3aOu5jpA0NNks5qMOb5gaJpZM4IRjAv
> .

kinakuta commented:
The problem isn't with clicking on the item and having it show up, but that's where the solution is to the real problem, which is reloading the browser once you've clicked on that item. Clicking on that item changes the url, so if you refresh the browser at that point, your Backbone router is then responsible for calling the same appController method clicking on the item ultimately calls (by way of firing an event.) The router, however, doesn't have any notion of what model the url represents, it only knows the id that appears on the url. Since it needs to pass the id to the method, that method needs to get an id, not a model. Clicking on the item does have access to the id, so it has the flexibility to change that the router does not. Does that make sense? The id is available on the model, so send it instead:

`vent.trigger('house:selected', { id: this.model.id, path: path });`

bkat70 commented:
hi James,

thank you for the additional explanation - I now understand the problem and
actually see it. I tried several fixes (didn't work) and have now rewritten
the code that uses 'id' and  use "street" instead which is analogous to
your use of "name" in the respective code snippets.
There is some improvement: the house is served up again when I refresh-
however, somehow I "lose" my buttons, images etc. The console shows the
following message:  "_Resource interpreted as Stylesheet but transferred
with MIME type text/html: "http://127.0.0.1:3000/houses/css/house.css
http://127.0.0.1:3000/houses/css/house.css_". (see attached
ConsoleLog.png)

Digging around it looks like all the returned MIME types at that point are
the wrong kind (they are all text.html which explains why the image isn't
showing up as an image or the buttons as buttons). Furthermore, it looks
like the express folder I have installed for the FoodTruckAPI contains an
express folder with far more subfolders than the one installed for my
project. Yours contains a file "types.json" but I can't find a similar file
in my express nodule folder (see attached expressFolders.png)? Both
package.json files look similar (mine contains 'path' and 'body-parser as
well), so I am at loss why my project would be missing those files. Any
idea(s)? I thought all those folders would install themselves....

again, thank you for your help!
katja

On Thu, Jun 16, 2016 at 10:45 AM, James Donaldson notifications@github.com
wrote:

> The problem isn't with clicking on the item and having it show up, but
> that's where the solution is to the real problem, which is reloading the
> browser once you've clicked on that item. Clicking on that item changes the
> url, so if you refresh the browser at that point, your Backbone router is
> then responsible for calling the same appController method clicking on the
> item ultimately calls (by way of firing an event.) The router, however,
> doesn't have any notion of what model the url represents, it only knows the
> id that appears on the url. Since it needs to pass the id to the method,
> that method needs to get an id, not a model. Clicking on the item does have
> access to the id, so it has the flexibility to change that the router does
> not. Does that make sense? The id is available on the model, so send it
> instead:
> 
> vent.trigger('house:selected', { id: this.model.id, path: path });
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/JSCRIPT300-Spring2016/Final_Project/pull/2#issuecomment-226560130,
> or mute the thread
> https://github.com/notifications/unsubscribe/ARbciL2CPUVPb2FXGF3bR1D1DvM90dguks5qMYvAgaJpZM4IRjAv
> .

bkat70 committed:
changed a couple of files

jamdon commented on public/index.html:
```diff
@@ -0,0 +1,32 @@
+<head>
+  <meta charset="UTF-8">
+  <title>Mock Real Estate Platform</title>
+  <link rel="stylesheet" href="css/house.css" />
```
This needs to be an absolute path from your root: `href="/css/house.css"`

The spa-server-route module is serving up index.html for any routes your express router doesn't explicitly support. But this means that if it's serving it up from something like `/houses/somestreetname`, the relative paths inside your index.html file will be resolved to `/houses/css/house.css` or whatever else the resource name is, and it won't be found by the express.static middleware in the public directory like it expects. Make any resource paths like this hard-coded as if they're from the root (I believe you have some image files with similar paths) and it should fix the way your pages look if you load the server from your `/houses/somestreetname` routes.

jamdon commented:
I left an inline comment explaining what's going on with how to fix it.

bkat70 commented:
ahhh, that did it - thanks!!!!

k.

On Thu, Jun 16, 2016 at 10:17 PM, jamdon notifications@github.com wrote:

> I left an inline comment explaining what's going on with how to fix it.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/JSCRIPT300-Spring2016/Final_Project/pull/2#issuecomment-226682488,
> or mute the thread
> https://github.com/notifications/unsubscribe/ARbciAt4TqZZ_Pa3ASHc6DGGmAzM72JVks5qMi3WgaJpZM4IRjAv
> .

bkat70 committed:
updated index.js and index.html

bkat70 commented:
Do I need to make another pull-request or does the open one suffice?

thanks!
katja

On Thu, Jun 16, 2016 at 10:29 PM, Katja Borchert katjab@uw.edu wrote:

> ahhh, that did it - thanks!!!!
> 
> k.
> 
> On Thu, Jun 16, 2016 at 10:17 PM, jamdon notifications@github.com wrote:
> 
> > I left an inline comment explaining what's going on with how to fix it.
> > 
> > —
> > You are receiving this because you authored the thread.
> > Reply to this email directly, view it on GitHub
> > https://github.com/JSCRIPT300-Spring2016/Final_Project/pull/2#issuecomment-226682488,
> > or mute the thread
> > https://github.com/notifications/unsubscribe/ARbciAt4TqZZ_Pa3ASHc6DGGmAzM72JVks5qMi3WgaJpZM4IRjAv
> > .

jamdon commented:
Works great! Great job persevering through this! 👍

bkat70 commented:
Thank YOU!

Hope you are enjoying your Monday evening - i am :)

Katja

On Thursday, June 16, 2016, jamdon notifications@github.com wrote:

> Works great! Great job persevering through this! 👍
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/JSCRIPT300-Spring2016/Final_Project/pull/2#issuecomment-226684625,
> or mute the thread
> https://github.com/notifications/unsubscribe/ARbciFgVwoqFRTdcl4TzbJ5XuuUTEn_2ks5qMjKLgaJpZM4IRjAv
> .
